src/middlewared/middlewared/plugins/activedirectory.py

```diff
@@ -0,0 +1,1147 @@
+                        'datastore.update',
+                        'directoryservice.kerberosrealm',
+                        ad['kerberos_realm']['id'],
+                        site_indexed_kerberos_data
+                    )
+                    await self.middleware.call('etc.generate', 'kerberos')
+
+        if smb['workgroup'] == 'WORKGROUP':
+            smb['workgroup'] = await asyncio.wait_for(self.get_netbios_domain_name(), 10)
+
+        await self.middleware.call('etc.generate', 'smb')
+
+        """
+        Check response of 'net ads testjoin' to determine whether the server needs to be joined to Active Directory.
+        Only perform the domain join if we receive the exact error code indicating that the server is not joined to
+        Active Directory. 'testjoin' will fail if the NAS boots before the domain controllers in the environment.
+        In this case, samba should be started, but the directory service reported in a FAULTED state.
+        """
+
+        ret = await self._net_ads_testjoin(smb['workgroup'])
+        if ret == neterr.NOTJOINED:
+            self.logger.debug(f"Test join to {ad['domainname']} failed. Performing domain join.")
+            await self._net_ads_join()
+            kt_set = await self.middleware.call('kerberos.keytab.store_samba_keytab')
+            if kt_set:
+                self.logger.debug('Successfully generated keytab for computer account. Clearing bind credentials')
+                await self.update({'bindpw': '', 'kerberos_principal': f'{smb["netbiosname"]}$@{ad["domainname"]}'})
+            ret = neterr.JOINED
+            if ad['allow_trusted_doms']:
+                await self.middleware.call('idmap.autodiscover_trusted_domains')
+
+        await self.middleware.call('service.restart', 'cifs')
+        await self.middleware.call('etc.generate', 'pam')
+        await self.middleware.call('etc.generate', 'nss')
+        if ret == neterr.JOINED:
+            await self._set_state(DSStatus['HEALTHY'])
+            await self.get_ad_cache()
+        else:
+            await self._set_state(DSStatus['FAULTED'])
+
+    @private
+    async def stop(self):
+        ad = await self.config()
+        await self.middleware.call('datastore.update', self._config.datastore, ad['id'], {'ad_enable': False})
+        await self._set_state(DSStatus['LEAVING'])
+        await self.middleware.call('etc.generate', 'hostname')
+        await self.middleware.call('kerberos.stop')
+        await self.middleware.call('etc.generate', 'smb')
+        await self.middleware.call('service.restart', 'cifs')
+        await self.middleware.call('etc.generate', 'pam')
+        await self.middleware.call('etc.generate', 'nss')
+        await self.middleware.call('cache.pop', 'AD_State')
+
+    @private
+    async def validate_credentials(self):
+        ret = False
+        ad = await self.config()
+        with ActiveDirectory_DNS(conf=ad, logger=self.logger) as AD_DNS:
```

Comment:
This dos synchronous calls, correct? If so `validate_credentials` should be a method, not a coroutine